arangod/Aql/OptimizerRules.cpp

```diff
@@ -1266,18 +1270,75 @@ void arangodb::aql::removeCollectVariablesRule(Optimizer* opt,
             auto usedThere =
                 Ast::getReferencedAttributesForKeep(exp->node(), outVariable,
                                                     isSafeForOptimization);
+
+
             if (isSafeForOptimization) {
               for (auto const& it : usedThere) {
                 keepAttributes.emplace(it);
               }
             } else {
               stop = true;
+              break;
+            }
+
+            if(other != nullptr) {
+
+              auto otherOutvar = other->outVariable();
+
+              // check if the name of the current variable turns up as attribute access in the next collect
+              bool isSafeForOptimization;
+              Ast::getReferencedAttributesForKeep(exp->node(), outVariable, isSafeForOptimization, true /*test indexed*/);
+              if(!isSafeForOptimization) {
+                stop = true;
+              }
+
+              // check if within the next collect the new group variable is used without any attribute
+              // if so the current variabe has to be available.
+              auto usedThere =
+                Ast::getReferencedAttributesForKeep(exp->node(), otherOutvar,
+                                                    isSafeForOptimization);
+              if(usedThere.empty()) {
+                stop = true;
+              }
+
+            } // end - other != nullptr
+          }
+        } else if (p->getType() == EN::RETURN && other) {
```

Comment:
Should this actually be restricted to just `RETURN`?
What about any other nodes that use the variable as their input?